src/main/java/org/fcrepo/migration/f4clients/StatelessFedora4Client.java

```diff
@@ -0,0 +1,237 @@
+    }
+
+    private void assertSuccess(final FcrepoResponse r) {
+        if (!success(r)) {
+            throw new RuntimeException("error code " + r.getStatusCode() + " from request " + r.getUrl());
+        }
+    }
+
+    private URI pathToURI(final String path) throws URISyntaxException {
+        return path.startsWith(baseUri) ? new URI(path) : new URI(baseUri + path);
+    }
+
+    private String uriToPath(final String URI) {
+        if (URI.startsWith(baseUri)) {
+            return URI.substring(baseUri.length());
+        } else {
+            return URI;
+        }
+    }
+
+    @Override
+    public boolean exists(final String path) {
+        try {
+            return getClient().head(pathToURI(path)).getStatusCode() != 404;
+        } catch (FcrepoOperationFailedException | URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void createResource(final String path) {
+        try {
+            assertSuccess(getClient().put(pathToURI(path), null, null));
+        } catch (FcrepoOperationFailedException | URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public String getRepositoryUrl() {
+        return baseUri.toString();
+    }
+
+    @Override
+    public void createOrUpdateRedirectNonRDFResource(final String path, final String url) {
+        try {
+            assertSuccess(getClient().put(pathToURI(path), null,
+                    "message/external-body; access-type=URL; URL=\"" + url.toString() + "\""));
+        } catch (FcrepoOperationFailedException | URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void createOrUpdateNonRDFResource(final String path, final InputStream content, final String contentType) {
+        try {
+            assertSuccess(getClient().put(pathToURI(path), content, contentType));
+        } catch (FcrepoOperationFailedException | URISyntaxException e) {
+            throw new RuntimeException(e);
+        } finally {
+            try {
+                content.close();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+    }
+
+    @Override
+    public void createVersionSnapshot(final String path,final  String versionId) {
+        try {
+            final FcrepoHttpClientBuilder client
+                    = new FcrepoHttpClientBuilder(username, password, new URI(baseUri).toURL().getHost());
+            final CloseableHttpClient c = client.build();
+            final HttpMethods method = HttpMethods.POST;
+            final URI uri = pathToURI(path + "/fcr:versions");
+            final HttpEntityEnclosingRequestBase request = (HttpEntityEnclosingRequestBase) method.createRequest(uri);
+            request.addHeader("Slug", versionId);
+            final HttpResponse response = c.execute(request);
```

Comment:
This is a `CloseableHttpResponse` and also belongs inside a `try-with-resource`.